Use Storages for Cudagraph liveness

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96775
* #96433
* #96384
* __->__ #96154
* #89146
* #95020
* #94943
* #94653

In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Differential Revision: [D43999890](https://our.internmc.facebook.com/intern/diff/D43999890)

eellison committed:
Use Storages for Cudagraph liveness

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999890).

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999890](https://our.internmc.facebook.com/intern/diff/D43999890)

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999890](https://our.internmc.facebook.com/intern/diff/D43999890)

[ghstack-poisoned]

eellison committed:
Update on "Use Storages for Cudagraph liveness"


In the previous PR we used tensors for liveness. This isn't actually correct as we should be treating a cudagraph output as dead only when its storage is dead, not the tensor impl output. The previous handling also had a particularly bad interaction with aot autograd detaching, since it would treat all outputs as dead. 

I'm using `StorageWeakRef` from `reductions.py`. This is sufficient for checking if a storage has died, although it doesn't allow you to trigger an event immediately upon deallocation. That works fine for most of the handling of the storage weak ref. The one case it doesn't is if you deallocated a torch dynamo function and its cudagraph wrapper, but you still have a Tensor output of it that you need to wait on to deallocate the cudagraph memory pool.  We would need to trigger an event in order to do proper deferred deletion of the cudagraph. I'm deferring that till https://github.com/pytorch/pytorch/issues/91395 has landed, but that shouldn't be a particularly common case. 



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999890](https://our.internmc.facebook.com/intern/diff/D43999890)

[ghstack-poisoned]

eellison commented:
Squashed into https://github.com/pytorch/pytorch/pull/89146
